aten/src/ATen/native/cpu/ReduceOpsKernel.cpp

```diff
@@ -136,15 +136,34 @@ static void logcumsumexp_cpu_kernel(Tensor& result, const Tensor& self, int64_t
   });
 }
 
-static void mean_kernel_impl(TensorIterator& iter) {
-  AT_DISPATCH_ALL_TYPES_AND_COMPLEX(iter.dtype(), "mean_cpu", [&] {
-    scalar_t factor = scalar_t(iter.num_output_elements()) / scalar_t(iter.numel());
-    binary_kernel_reduce(
+template <typename scalar_t, typename acc_t=scalar_t, typename out_t=scalar_t>
+void mean_kernel_impl(TensorIterator& iter) {
+  //  returns acc_t for all non-complex dtypes and returns T for c10::complex<T>
+  using factor_t = typename c10::scalar_value_type<acc_t>::type;
+  factor_t factor = static_cast<factor_t>(iter.num_output_elements()) / iter.numel();
+  binary_kernel_reduce(
       iter,
-      MeanOps<scalar_t, scalar_t> {factor},
-      scalar_t(0)
+      MeanOps<scalar_t, acc_t, factor_t, out_t> {factor},
+      acc_t(0)
     );
-  });
+}
+
+static void mean_kernel_cpu(TensorIterator& iter) {
+  if (iter.dtype() == kHalf) {
+    mean_kernel_impl<at::Half, float>(iter);
+  } else if (iter.dtype(1) == kHalf && iter.dtype() == kFloat) {
+    // type promotion that does cast and reduction in a single kernel
+    mean_kernel_impl<at::Half, float, float>(iter);
```

Comment:
This special case only exists for CUDA, see `make_reduction_from_out_ty`:
https://github.com/pytorch/pytorch/blob/364d92f9b6864ce284fa13519c7ca5c87460e477/aten/src/ATen/native/ReduceOpsUtils.h#L434-L442

Reply:
removed.